File: polyglot/app_yaml_overwrites/mjs/src/context-builder.ts
/**
 * Generic request type for framework-agnostic usage.
 * Compatible with Fastify, Express, or any framework with headers.
 */
export interface RequestLike {
    headers?: Record<string, string | string[] | undefined>;
    [key: string]: any;
}

export interface ContextOptions {
    env?: Record<string, string>;
    config?: any; // AppYamlConfig raw
    app?: any;    // Extracted app config
    state?: any;
    request?: RequestLike;
}

export type ContextExtender = (currentContext: any, request?: RequestLike) => Promise<any> | any;

export class ContextBuilder {
    static async build(options: ContextOptions, extenders: ContextExtender[] = []): Promise<any> {
        const baseContext = {
            env: options.env || process.env,
            config: options.config || {},
            app: options.app || {},
            state: options.state || {},
            request: options.request, // Only present if request scope
        };

        let context = { ...baseContext };

        for (const extender of extenders) {
            const partial = await extender(context, options.request);
            context = { ...context, ...partial };
        }

        return context;
    }
}

File: polyglot/app_yaml_overwrites/py/src/app_yaml_overwrites/context_builder.py
import os
from typing import Dict, Any, Optional, List, Callable, Awaitable

ContextExtender = Callable[[Dict[str, Any], Optional[Any]], Awaitable[Dict[str, Any]]]

class ContextBuilder:
    @staticmethod
    async def build(
        options: Dict[str, Any], 
        extenders: List[ContextExtender] = None
    ) -> Dict[str, Any]:
        """
        Builds the resolution context.
        options: dict containing 'env', 'config', 'app', 'state', 'request'
        extenders: list of async functions to extend context
        """
        request = options.get("request")
        
        base_context = {
            "env": options.get("env", dict(os.environ)),
            "config": options.get("config", {}),
            "app": options.get("app", {}),
            "state": options.get("state", {}),
            "request": request,
        }
        
        context = base_context.copy()
        
        if extenders:
            for extender in extenders:
                partial = await extender(context, request)
                context.update(partial)
                
        return context
